lib/mpl_toolkits/mplot3d/art3d.py

```diff
@@ -229,23 +245,33 @@ def set_segments(self, segments):
         """
         Set 3D segments.
         """
-        self._segments3d = np.asanyarray(segments)
+        self._segments3d = []
+        if len(segments) > 0:
+            self._seg_sizes = [len(c) for c in segments]
+            # Store the points in a single array for easier projection
+            n_segments = np.sum(self._seg_sizes)
+
+            self._segments3d_data = np.empty((n_segments, 4))
+            self._segments3d_data[:, :3] = np.vstack(segments)
+            self._segments3d_data[:, 3] = 1
```

Comment:
https://github.com/matplotlib/matplotlib/pull/6085#discussion_r128392253 by @eric-wieser 

> This would be marginally better as
> 
> ```py
> self._segments3d_data  = np.empty((n_segments, 4))
> self._segments3d_data[:,:3] = segments
> self._segments3d_data[:,3] = 1
> ```

Done.

> Also, are you sure you mean "quaternion" and not "homogenous coordinate"?

Not sure, but I removed the "quaternion" term as suggested.